src/System.Security.Permissions/src/System/Security/SecurityElement.cs

```diff
@@ -2,30 +2,827 @@
+            SecurityElement element = new SecurityElement(m_strTag, m_strText);
+            element.m_lChildren = m_lChildren == null ? null : new ArrayList(m_lChildren);
+            element.m_lAttributes = m_lAttributes == null ? null : new ArrayList(m_lAttributes);
+
+            return element;
+        }
+
+        public static bool IsValidTag(string tag)
+        {
+            if (tag == null)
+                return false;
+
+            return tag.IndexOfAny(s_tagIllegalCharacters) == -1;
+        }
+
+        public static bool IsValidText(string text)
+        {
+            if (text == null)
+                return false;
+
+            return text.IndexOfAny(s_textIllegalCharacters) == -1;
+        }
+
+        public static bool IsValidAttributeName(string name)
+        {
+            return IsValidTag(name);
+        }
+
+        public static bool IsValidAttributeValue(string value)
+        {
+            if (value == null)
+                return false;
+
+            return value.IndexOfAny(s_valueIllegalCharacters) == -1;
+        }
+
+        private static string GetEscapeSequence(char c)
+        {
+            int iMax = s_escapeStringPairs.Length;
+            Debug.Assert(iMax % 2 == 0, "Odd number of strings means the attr/value pairs were not added correctly");
+
+            for (int i = 0; i < iMax; i += 2)
+            {
+                string strEscSeq = s_escapeStringPairs[i];
+                string strEscValue = s_escapeStringPairs[i + 1];
+
+                if (strEscSeq[0] == c)
+                    return strEscValue;
+            }
+
+            Debug.Assert(false, "Unable to find escape sequence for this character");
+            return c.ToString();
+        }
+
+        public static string Escape(string str)
+        {
+            if (str == null)
+                return null;
+
+            StringBuilder sb = null;
+
+            int strLen = str.Length;
+            int index; // Pointer into the string that indicates the location of the current '&' character
+            int newIndex = 0; // Pointer into the string that indicates the start index of the "remaining" string (that still needs to be processed).
+
+            do
+            {
+                index = str.IndexOfAny(s_escapeChars, newIndex);
+
+                if (index == -1)
+                {
+                    if (sb == null)
+                        return str;
+                    else
+                    {
+                        sb.Append(str, newIndex, strLen - newIndex);
+                        return sb.ToString();
+                    }
+                }
+                else
+                {
+                    if (sb == null)
+                        sb = new StringBuilder();
+
+                    sb.Append(str, newIndex, index - newIndex);
+                    sb.Append(GetEscapeSequence(str[index]));
+
+                    newIndex = (index + 1);
+                }
+            }
+            while (true);
```

Comment:
Personally, do..while usually suggests the first iteration is unusual; do..while(true) should probably just be a while.